### PR TITLE
shader_decompiler: Implemented (Partialy) Control Codes and CSETP

### PR DESCRIPTION
**Pull request by FernandoS27**
After some intense RE, we were able to get the behavior of CSETP and Control Codes. This PR only implements the base Control Codes, CSETP and an example of their use, as many instructions actually modify them.

### CODE REVIEW

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 1]
```diff
@@ -2014,31 +2033,55 @@ class GLSLGenerator {
...
+                if (instr.psetp.pred0 != static_cast<u64>(Pred::UnusedIndex)) {
+                    // Set the secondary predicate to the result of !Predicate OP SecondPredicate,
+                    // if enabled
+                    SetPredicate(instr.psetp.pred0,
+                                 "!(" + predicate + ") " + combiner + " (" + second_pred + ')');
+                }
+                break;
+            }
+            case OpCode::Id::CSETP: {
+                std::string pred =
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: These can be const.

File: src/video_core/engines/shader_bytecode.h [hunk 2]
```diff
@@ -984,6 +1030,7 @@ class OpCode {
             INST("010010110101----", Id::ISET_C, Type::IntegerSet, "ISET_C"),
             INST("0011011-0101----", Id::ISET_IMM, Type::IntegerSet, "ISET_IMM"),
             INST("0101000010010---", Id::PSETP, Type::PredicateSetPredicate, "PSETP"),
+            INST("010100001010----", Id::PSETP, Type::PredicateSetPredicate, "CSETP"),
```
Comment (Subv) [inline src/video_core/engines/shader_bytecode.h]: Wrong opcode

Comment (Hexagon12) [inline src/video_core/engines/shader_bytecode.h]: *Id::CSETP*

Comment (FernandoS27) [reply src/video_core/engines/shader_bytecode.h]: Control Codes is indeed a 32 bits flag register. I implemented it using 32 1 bit variables to abuse GLSL compiler's deadcode elimination and thus don't affect performance so heavily. As of right now this implementation does not implement everything, there's still a bunch of flags to implement. i just wanted to write a single example on how it is done.

I'll amend the title to signify it's partial.

**Comment by FernandoS27** [comment]
Changes made.

**Comment by FernandoS27** [comment]
Rebased

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 3]
```diff
@@ -1576,6 +1591,11 @@ class GLSLGenerator {
 
                 regs.SetRegisterToInteger(instr.gpr0, instr.conversion.is_output_signed, 0, op_a, 1,
                                           1, instr.alu.saturate_d, 0, instr.conversion.dest_size);
+                if (instr.generates_cc.Value() != 0) {
+                    const std::string neucondition =
+                        "( " + regs.GetRegisterAsInteger(instr.gpr0, 0, false) + " != 0 )";
```
[the remark this answers is not in the document]
Comment (FernandoS27) [reply src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: You are right

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 4]
```diff
@@ -1641,6 +1673,10 @@ class GLSLGenerator {
 
                 regs.SetRegisterToInteger(instr.gpr0, instr.conversion.is_output_signed, 0, op_a, 1,
                                           1, instr.alu.saturate_d, 0, instr.conversion.dest_size);
+                if (instr.generates_cc.Value() != 0) {
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: I think this code might be wrong, it doesn't take into account the saturation stuff nor the destination size.

I think it's better if you pass generates_cc to SetRegister and do the flag setting there.

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 5]
```diff
@@ -351,6 +359,24 @@ class GLSLRegisterManager {
         shader.AddLine(dest + " = " + src + ';');
     }
 
+    std::string GetControlCode(const Tegra::Shader::ControlCode cc) const {
+        switch (cc) {
+        case Tegra::Shader::ControlCode::NEU:
+            return "!(" + GetInternalFlag(InternalFlag::ZeroFlag) + ')';
+        default:
```
Comment (Subv) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: log + UNREACHABLE in the default case.

**Comment by FernandoS27** [comment]
that shouldn't be a problem if the IR is SSA. It's a piece of cake to eliminate deadcode in that format.

**Comment by bunnei** [comment]
Thanks!

**Pull request merged**